File: app/src/main/java/ru/p3tr0vich/fuel/PreferencesHelper.java
package ru.p3tr0vich.fuel;

import android.annotation.SuppressLint;
import android.content.ContentValues;
import android.content.Context;
import android.content.SharedPreferences;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.preference.PreferenceManager;
import android.support.annotation.IntDef;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.util.Map;

class PreferencesHelper {

    private static final String TAG = "PreferencesHelper";

    public static final long SYNC_NONE = Long.MIN_VALUE;

    public static final String PREF_DATABASE_REVISION = "database revision";
    public static final String PREF_PREFERENCES_REVISION = "preferences revision";

    public static final String PREF_CHANGED = "changed";
    public static final String PREF_LAST_SYNC_DATE_TIME = "last sync date time";
    public static final String PREF_LAST_SYNC_HAS_ERROR = "last sync has error";
    public static final String PREF_DATABASE_FULL_SYNC = "database full sync";

    private static final String PREF_FILTER_DATE_FROM = "filter date from";
    private static final String PREF_FILTER_DATE_TO = "filter date to";

    private static final String PREF_DISTANCE = "distance";
    private static final String PREF_COST = "cost";
    private static final String PREF_VOLUME = "volume";
    private static final String PREF_PRICE = "price";

    private static final String PREF_CONS = "consumption";
    private static final String PREF_SEASON = "season";

    // @preferences
    public static final String PREF_MAP_CENTER_TEXT = "map center text";

    private static final String PREF_MAP_CENTER_LATITUDE = "map center latitude";
    private static final String PREF_MAP_CENTER_LONGITUDE = "map center longitude";

    public static final String DEFAULT_MAP_CENTER_TEXT = "Москва, Кремль";
    public static final double DEFAULT_MAP_CENTER_LATITUDE = 55.752023;  // Широта
    public static final double DEFAULT_MAP_CENTER_LONGITUDE = 37.617499; // Долгота

    // @preferences
    public static final String PREF_SYNC = "key_sync";
    public static final String PREF_SYNC_ENABLED = "sync enabled";
    public static final String PREF_SYNC_YANDEX_DISK = "key_sync_yandex_disk";

    public static final String PREF_SMS = "key_sms";
    public static final String PREF_SMS_ENABLED = "sms enabled";
    public static final String PREF_SMS_ADDRESS = "sms address";
    public static final String PREF_SMS_TEXT = "sms text";

    @SuppressWarnings("WeakerAccess")
    // private - поле удаляется сборщиком мусора
    static SharedPreferences.OnSharedPreferenceChangeListener sPreferenceChangeListener; // TODO: this
    private static Context sContext;
    private static SharedPreferences sSharedPreferences;

    @Retention(RetentionPolicy.SOURCE)
    @IntDef({PREFERENCE_TYPE_STRING, PREFERENCE_TYPE_INT, PREFERENCE_TYPE_LONG})
    public @interface PreferenceType {
    }

    public static final int PREFERENCE_TYPE_STRING = 0;
    public static final int PREFERENCE_TYPE_INT = 1;
    public static final int PREFERENCE_TYPE_LONG = 2;

    private PreferencesHelper() {

    }

    public static void init(@NonNull Context context) {
        sContext = context;

        sSharedPreferences = PreferenceManager.getDefaultSharedPreferences(sContext);

//        sSharedPreferences.edit()
//                .remove("default cost")
//                .apply();

        sPreferenceChangeListener = new SharedPreferences.OnSharedPreferenceChangeListener() {
            @Override
            public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
                UtilsLog.d(TAG, "onSharedPreferenceChanged", "key == " + key);

                if (isSyncKey(key)) putChanged(true);
            }
        };
        sSharedPreferences.registerOnSharedPreferenceChangeListener(sPreferenceChangeListener);
    }

    private static boolean isSyncKey(String key) {
        switch (key) {
            case PREF_CHANGED:
            case PREF_DATABASE_REVISION:
            case PREF_PREFERENCES_REVISION:
            case PREF_LAST_SYNC_DATE_TIME:
            case PREF_LAST_SYNC_HAS_ERROR:
            case PREF_DATABASE_FULL_SYNC:
            case PREF_SYNC_ENABLED:
            case PREF_SMS_ENABLED:
                return false;
            default:
                return true;
        }
    }

    private static boolean isChanged() {
        return sSharedPreferences.getBoolean(PREF_CHANGED, true);
    }

    private static void putChanged(final boolean changed) {
        sSharedPreferences
                .edit()
                .putBoolean(PREF_CHANGED, changed)
                .apply();

        UtilsLog.d(TAG, "putChanged", "changed == " + changed);

        if (changed)
            sContext.getContentResolver().notifyChange(ContentProviderHelper.URI_PREFERENCES, null, false);
    }

    public static boolean isSyncEnabled() {
        return sSharedPreferences.getBoolean(PREF_SYNC_ENABLED, false);
    }

    public static boolean isSMSEnabled() {
        return sSharedPreferences.getBoolean(PREF_SMS_ENABLED, false);
    }

    @NonNull
    public static String getSMSAddress() {
        return getString(PREF_SMS_ADDRESS);
    }

    public static void putSMSAddress(final String address) {
        sSharedPreferences
                .edit()
                .putString(PREF_SMS_ADDRESS, address)
                .apply();
    }

    @NonNull
    public static String getSMSText() {
        return getString(PREF_SMS_TEXT);
    }

    private static int getRevision(String keyRevision) {
        return sSharedPreferences.getInt(keyRevision, -1);
    }

    private static void putRevision(final String keyRevision, final int revision) {
        sSharedPreferences
                .edit()
                .putInt(keyRevision, revision)
                .apply();

        UtilsLog.d(TAG, "putRevision", keyRevision + " == " + revision);
    }

    private static boolean isFullSync() {
        return sSharedPreferences.getBoolean(PREF_DATABASE_FULL_SYNC, false);
    }

    public static void putFullSync(final boolean fullSync) {
        sSharedPreferences
                .edit()
                .putBoolean(PREF_DATABASE_FULL_SYNC, fullSync)
                .apply();
    }

    public static long getLastSyncDateTime() {
        return sSharedPreferences.getLong(PREF_LAST_SYNC_DATE_TIME, SYNC_NONE);
    }

    public static boolean getLastSyncHasError() {
        return sSharedPreferences.getBoolean(PREF_LAST_SYNC_HAS_ERROR, false);
    }

    private static void putLastSyncDateTime(final long dateTime) {
        sSharedPreferences
                .edit()
                .putLong(PREF_LAST_SYNC_DATE_TIME, dateTime)
                .apply();
    }

    private static void putLastSyncHasError(final boolean hasError) {
        sSharedPreferences
                .edit()
                .putBoolean(PREF_LAST_SYNC_HAS_ERROR, hasError)
                .apply();
    }

    public static long getFilterDateFrom() {
        return sSharedPreferences.getLong(PREF_FILTER_DATE_FROM, System.currentTimeMillis());
    }

    public static long getFilterDateTo() {
        return sSharedPreferences.getLong(PREF_FILTER_DATE_TO, System.currentTimeMillis());
    }

    public static void putFilterDate(final long dateFrom, final long dateTo) {
        sSharedPreferences
                .edit()
                .putLong(PREF_FILTER_DATE_FROM, dateFrom)
                .putLong(PREF_FILTER_DATE_TO, dateTo)
                .apply();
    }

    public static float getDefaultCost() {
        return UtilsFormat.stringToFloat(getString(sContext.getString(R.string.pref_def_cost)));
    }

    public static float getDefaultVolume() {
        return UtilsFormat.stringToFloat(getString(sContext.getString(R.string.pref_def_volume)));
    }

    public static float getLastTotal() {
        return UtilsFormat.stringToFloat(getString(sContext.getString(R.string.pref_last_total)));
    }

    public static void putLastTotal(final float lastTotal) {
        sSharedPreferences
                .edit()
                .putString(sContext.getString(R.string.pref_last_total), String.valueOf(lastTotal))
                .apply();
    }

    @NonNull
    public static String getCalcDistance() {
        return getString(PREF_DISTANCE);
    }

    @NonNull
    public static String getCalcCost() {
        return getString(PREF_COST);
    }

    @NonNull
    public static String getCalcVolume() {
        return getString(PREF_VOLUME);
    }

    @NonNull
    public static String getCalcPrice() {
        return getString(PREF_PRICE);
    }

    public static float[][] getCalcCons() {

        float[][] result = {{0, 0, 0}, {0, 0, 0}};

        result[0][0] = UtilsFormat.stringToFloat(getString(sContext.getString(R.string.pref_summer_city)));
        result[0][1] = UtilsFormat.stringToFloat(getString(sContext.getString(R.string.pref_summer_highway)));
        result[0][2] = UtilsFormat.stringToFloat(getString(sContext.getString(R.string.pref_summer_mixed)));
        result[1][0] = UtilsFormat.stringToFloat(getString(sContext.getString(R.string.pref_winter_city)));
        result[1][1] = UtilsFormat.stringToFloat(getString(sContext.getString(R.string.pref_winter_highway)));
        result[1][2] = UtilsFormat.stringToFloat(getString(sContext.getString(R.string.pref_winter_mixed)));

        return result;
    }

    public static int getCalcSelectedCons() {
        return sSharedPreferences.getInt(PREF_CONS, 0);
    }

    public static int getCalcSelectedSeason() {
        return sSharedPreferences.getInt(PREF_SEASON, 0);
    }

    public static void putCalc(final String distance, final String cost, final String volume,
                               final String price, final int cons, final int season) {
        sSharedPreferences
                .edit()
                .putString(PREF_DISTANCE, distance)
                .putString(PREF_COST, cost)
                .putString(PREF_VOLUME, volume)
                .putString(PREF_PRICE, price)
                .putInt(PREF_CONS, cons)
                .putInt(PREF_SEASON, season)
                .apply();
    }

    @NonNull
    public static String getMapCenterText() {
        return getString(PREF_MAP_CENTER_TEXT, DEFAULT_MAP_CENTER_TEXT);
    }

    public static double getMapCenterLatitude() {
        return Double.longBitsToDouble(sSharedPreferences.getLong(
                PREF_MAP_CENTER_LATITUDE,
                Double.doubleToLongBits(DEFAULT_MAP_CENTER_LATITUDE)));
    }

    public static double getMapCenterLongitude() {
        return Double.longBitsToDouble(sSharedPreferences.getLong(
                PREF_MAP_CENTER_LONGITUDE,
                Double.doubleToLongBits(DEFAULT_MAP_CENTER_LONGITUDE)));
    }

    public static void putMapCenter(final String text, final double latitude, final double longitude) {
        sSharedPreferences
                .edit()
                .putString(PREF_MAP_CENTER_TEXT, text)
                .putLong(PREF_MAP_CENTER_LATITUDE,
                        Double.doubleToRawLongBits(latitude))
                .putLong(PREF_MAP_CENTER_LONGITUDE,
                        Double.doubleToRawLongBits(longitude))
                .apply();
    }

    @NonNull
    private static String getString(final String key, final String defValue) {
        return sSharedPreferences.getString(key, defValue);
    }

    @NonNull
    public static String getString(final String key) {
        return getString(key, "");
    }

    @NonNull
    private static ContentValues getPreferences(@Nullable String preference) {
        ContentValues result = new ContentValues();

        if (TextUtils.isEmpty(preference)) {
            Map<String, ?> map = sSharedPreferences.getAll();

            Object value;

            for (String key : map.keySet())
                if (isSyncKey(key)) {

                    value = map.get(key);

                    if (value instanceof String) result.put(key, (String) value);
                    else if (value instanceof Long) result.put(key, (Long) value);
                    else if (value instanceof Integer) result.put(key, (Integer) value);
                    else if (value instanceof Boolean) result.put(key, (Boolean) value);
                    else if (value instanceof Float) result.put(key, (Float) value);
                    else
                        UtilsLog.d(TAG, "getPreferences",
                                "unhandled class == " + value.getClass().getSimpleName());
                }
        } else
            switch (preference) {
                case PREF_CHANGED:
                    result.put(preference, isChanged());
                    break;
                case PREF_DATABASE_FULL_SYNC:
                    result.put(preference, isFullSync());
                    break;
                case PREF_DATABASE_REVISION:
                case PREF_PREFERENCES_REVISION:
                    result.put(preference, getRevision(preference));
                    break;
                case PREF_LAST_SYNC_DATE_TIME:
                    result.put(preference, getLastSyncDateTime());
                    break;
                case PREF_LAST_SYNC_HAS_ERROR:
                    result.put(preference, getLastSyncHasError());
                    break;
                default:
                    UtilsLog.d(TAG, "getPreferences", "unhandled preference == " + preference);
            }

//        for (String key : result.keySet())
//            UtilsLog.d(TAG, "getPreferences", "key == " + key + ", value == " + result.getAsString(key));

        return result;
    }

    @NonNull
    private static Cursor getPreferencesCursor(@Nullable String preference) {
        MatrixCursor matrixCursor = new MatrixCursor(new String[]{"key", "value"});

        ContentValues preferences = PreferencesHelper.getPreferences(preference);
        for (String key : preferences.keySet())
            matrixCursor.addRow(new Object[]{key, preferences.get(key)});

        return matrixCursor;
    }

    @NonNull
    public static Cursor getPreferences() {
        return getPreferencesCursor(null);
    }

    @NonNull
    public static Cursor getPreference(@Nullable String preference) {
        return getPreferencesCursor(preference);
    }

    @SuppressLint("CommitPrefEdits")
    public static int setPreferences(@Nullable ContentValues preferences,
                                     @Nullable String preference) {
        if (preferences == null || preferences.size() == 0) return -1;

        UtilsLog.d(TAG, "setPreferences", "preference == " + preference);

        if (TextUtils.isEmpty(preference)) {
            sSharedPreferences.unregisterOnSharedPreferenceChangeListener(sPreferenceChangeListener);
            SharedPreferences.Editor editor = sSharedPreferences.edit();
            try {
                Object value;

                for (String key : preferences.keySet()) {
                    value = preferences.get(key);

                    if (value instanceof String) editor.putString(key, (String) value);
                    else if (value instanceof Long) editor.putLong(key, (Long) value);
                    else if (value instanceof Integer) editor.putInt(key, (Integer) value);
                    else if (value instanceof Boolean) editor.putBoolean(key, (Boolean) value);
                    else if (value instanceof Float) editor.putFloat(key, (Float) value);
                    else
                        UtilsLog.d(TAG, "setPreferences",
                                "unhandled class == " + value.getClass().getSimpleName());
                }
            } finally {
                editor.commit();
                sSharedPreferences.registerOnSharedPreferenceChangeListener(sPreferenceChangeListener);
            }

            return preferences.size();
        } else {
            switch (preference) {
                case PREF_CHANGED:
                    putChanged(preferences.getAsBoolean(preference));
                    break;
                case PREF_DATABASE_REVISION:
                case PREF_PREFERENCES_REVISION:
                    putRevision(preference, preferences.getAsInteger(preference));
                    break;
                case PREF_DATABASE_FULL_SYNC:
                    putFullSync(preferences.getAsBoolean(preference));
                    break;
                case PREF_LAST_SYNC_DATE_TIME:
                    putLastSyncDateTime(preferences.getAsLong(preference));
                    break;
                case PREF_LAST_SYNC_HAS_ERROR:
                    putLastSyncHasError(preferences.getAsBoolean(preference));
                    break;
                default:
                    UtilsLog.d(TAG, "setPreferences", "unhandled preference == " + preference);
            }

            return 1;
        }
    }

    @PreferenceType
    public static int getPreferenceType(@NonNull String key) {
        switch (key) {
            case PREF_CONS:
            case PREF_SEASON:
                return PREFERENCE_TYPE_INT;
            case PREF_FILTER_DATE_FROM:
            case PREF_FILTER_DATE_TO:
            case PREF_MAP_CENTER_LATITUDE:
            case PREF_MAP_CENTER_LONGITUDE:
                return PREFERENCE_TYPE_LONG;
            default:
                return PREFERENCE_TYPE_STRING;
        }
    }
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/helpers/PreferencesHelper.kt
package ru.p3tr0vich.fuel.helpers

import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.Context
import android.content.SharedPreferences
import android.database.Cursor
import android.database.MatrixCursor
import androidx.preference.PreferenceManager
import android.text.TextUtils
import androidx.annotation.IntDef
import ru.p3tr0vich.fuel.R
import ru.p3tr0vich.fuel.models.MapCenter
import ru.p3tr0vich.fuel.models.MapCenter.Companion.DEFAULT_MAP_CENTER_LATITUDE
import ru.p3tr0vich.fuel.models.MapCenter.Companion.DEFAULT_MAP_CENTER_LONGITUDE
import ru.p3tr0vich.fuel.models.MapCenter.Companion.DEFAULT_MAP_CENTER_TEXT
import ru.p3tr0vich.fuel.utils.UtilsFormat
import ru.p3tr0vich.fuel.utils.UtilsLog

class PreferencesHelper private constructor(private val context: Context) :
    SharedPreferences.OnSharedPreferenceChangeListener {

    private val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)

    val keys = Keys(context)

    private var isChanged: Boolean
        get() = sharedPreferences.getBoolean(keys.changed, true)
        set(value) {
            sharedPreferences
                .edit()
                .putBoolean(keys.changed, value)
                .apply()

            if (LOG_ENABLED) {
                UtilsLog.d(TAG, "putChanged", "value == $value")
            }

            if (value) {
                context.contentResolver.notifyChange(
                    ContentProviderHelper.URI_PREFERENCES,
                    null
                )
            }
        }

    val isSyncEnabled: Boolean
        get() = sharedPreferences.getBoolean(keys.syncEnabled, false)

    val isSMSEnabled: Boolean
        get() = sharedPreferences.getBoolean(keys.smsEnabled, false)

    var smsAddress: String
        get() = getString(keys.smsAddress)
        set(value) {
            sharedPreferences
                .edit()
                .putString(keys.smsAddress, value)
                .apply()
        }

    var smsTextPattern: String
        get() = getString(keys.smsTextPattern)
        set(value) {
            sharedPreferences
                .edit()
                .putString(keys.smsTextPattern, value)
                .apply()
        }

    var smsText: String
        get() = getString(keys.smsText)
        set(value) {
            sharedPreferences
                .edit()
                .putString(keys.smsText, value)
                .apply()
        }

    var isFullSync: Boolean
        get() = sharedPreferences.getBoolean(keys.databaseFullSync, false)
        set(value) {
            sharedPreferences
                .edit()
                .putBoolean(keys.databaseFullSync, value)
                .apply()
        }

    var lastSyncDateTime: Long
        get() = sharedPreferences.getLong(keys.lastSyncDateTime, SYNC_NONE)
        set(value) {
            sharedPreferences
                .edit()
                .putLong(keys.lastSyncDateTime, value)
                .apply()
        }

    var lastSyncHasError: Boolean
        get() = sharedPreferences.getBoolean(keys.lastSyncHasError, false)
        set(value) {
            sharedPreferences
                .edit()
                .putBoolean(keys.lastSyncHasError, value)
                .apply()
        }

    val filterDateFrom: Long
        get() = sharedPreferences.getLong(keys.filterDateFrom, System.currentTimeMillis())

    val filterDateTo: Long
        get() = sharedPreferences.getLong(keys.filterDateTo, System.currentTimeMillis())

    val defaultCost: Float
        get() = UtilsFormat.stringToFloat(getString(keys.defaultCost))

    val defaultVolume: Float
        get() = UtilsFormat.stringToFloat(getString(keys.defaultVolume))

    var lastTotal: Float
        get() = UtilsFormat.stringToFloat(getString(context.getString(R.string.pref_key_last_total)))
        set(value) {
            sharedPreferences
                .edit()
                .putString(context.getString(R.string.pref_key_last_total), value.toString())
                .apply()
        }

    val calcDistance: String
        get() = getString(keys.distance)

    val calcCost: String
        get() = getString(keys.cost)

    val calcVolume: String
        get() = getString(keys.volume)

    val priceAsString: String
        get() = getString(keys.price)

    val price: Float
        get() = UtilsFormat.stringToFloat(priceAsString)

    val calcCons: Array<FloatArray>
        get() {
            val result = arrayOf(floatArrayOf(0f, 0f, 0f), floatArrayOf(0f, 0f, 0f))

            result[0][0] =
                UtilsFormat.stringToFloat(getString(keys.summerCity))
            result[0][1] =
                UtilsFormat.stringToFloat(getString(keys.summerHighway))
            result[0][2] =
                UtilsFormat.stringToFloat(getString(keys.summerMixed))
            result[1][0] =
                UtilsFormat.stringToFloat(getString(keys.winterCity))
            result[1][1] =
                UtilsFormat.stringToFloat(getString(keys.winterHighway))
            result[1][2] =
                UtilsFormat.stringToFloat(getString(keys.winterMixed))

            return result
        }

    val calcSelectedCons: Int
        get() = sharedPreferences.getInt(keys.consumption, 0)

    val calcSelectedSeason: Int
        get() = sharedPreferences.getInt(keys.season, 0)

    var mapCenter: MapCenter
        get() = MapCenter(
            getString(keys.mapCenterText, DEFAULT_MAP_CENTER_TEXT),
            java.lang.Double.longBitsToDouble(
                sharedPreferences.getLong(
                    keys.mapCenterLatitude,
                    java.lang.Double.doubleToLongBits(DEFAULT_MAP_CENTER_LATITUDE)
                )
            ),
            java.lang.Double.longBitsToDouble(
                sharedPreferences.getLong(
                    keys.mapCenterLongitude,
                    java.lang.Double.doubleToLongBits(DEFAULT_MAP_CENTER_LONGITUDE)
                )
            )
        )
        set(value) {
            sharedPreferences
                .edit()
                .putString(keys.mapCenterText, value.text)
                .putLong(
                    keys.mapCenterLatitude,
                    java.lang.Double.doubleToRawLongBits(value.latitude)
                )
                .putLong(
                    keys.mapCenterLongitude,
                    java.lang.Double.doubleToRawLongBits(value.longitude)
                )
                .apply()
        }

    @Retention(AnnotationRetention.SOURCE)
    @IntDef(PREFERENCE_TYPE_STRING, PREFERENCE_TYPE_INT, PREFERENCE_TYPE_LONG)
    annotation class PreferenceType

    class Keys constructor(context: Context) {
        val price = context.getString(R.string.pref_key_price)

        val defaultCost = context.getString(R.string.pref_key_def_cost)
        val defaultVolume = context.getString(R.string.pref_key_def_volume)

        val mapCenterText = context.getString(R.string.pref_key_map_center_text)
        val mapCenterLatitude = context.getString(R.string.pref_key_map_center_latitude)
        val mapCenterLongitude = context.getString(R.string.pref_key_map_center_longitude)

        val sync = context.getString(R.string.pref_key_sync)
        val syncEnabled = context.getString(R.string.pref_key_sync_enabled)
        val syncYandexDisk = context.getString(R.string.pref_key_sync_yandex_disk)

        val sms = context.getString(R.string.pref_key_sms)
        val smsEnabled = context.getString(R.string.pref_key_sms_enabled)
        val smsAddress = context.getString(R.string.pref_key_sms_address)
        val smsText = context.getString(R.string.pref_key_sms_text)
        val smsTextPattern = context.getString(R.string.pref_key_sms_text_pattern)

        val databaseRevision = context.getString(R.string.pref_key_database_revision)
        val preferencesRevision = context.getString(R.string.pref_key_preferences_revision)

        val changed = context.getString(R.string.pref_key_changed)
        val lastSyncDateTime = context.getString(R.string.pref_key_last_sync_date_time)
        val lastSyncHasError = context.getString(R.string.pref_key_last_sync_has_error)
        val databaseFullSync = context.getString(R.string.pref_key_database_full_sync)

        val filterDateFrom = context.getString(R.string.pref_key_filter_date_from)
        val filterDateTo = context.getString(R.string.pref_key_filter_date_to)

        val distance = context.getString(R.string.pref_key_distance)
        val cost = context.getString(R.string.pref_key_cost)
        val volume = context.getString(R.string.pref_key_volume)

        val consumption = context.getString(R.string.pref_key_consumption)
        val season = context.getString(R.string.pref_key_season)

        val summerCity = context.getString(R.string.pref_key_summer_city)
        val summerHighway = context.getString(R.string.pref_key_summer_highway)
        val summerMixed = context.getString(R.string.pref_key_summer_mixed)
        val winterCity = context.getString(R.string.pref_key_winter_city)
        val winterHighway = context.getString(R.string.pref_key_winter_highway)
        val winterMixed = context.getString(R.string.pref_key_summer_mixed)

        fun isSyncKey(key: String?): Boolean {
            return when (key) {
                syncEnabled,
                smsEnabled,
                changed, databaseRevision, preferencesRevision,
                lastSyncDateTime, lastSyncHasError, databaseFullSync -> false

                consumption, cost, defaultCost, defaultVolume, distance -> true

                filterDateFrom, filterDateTo -> true

                mapCenterLatitude, mapCenterLongitude, mapCenterText -> true

                price, season -> true
                summerCity, summerHighway, summerMixed,
                winterCity, winterHighway, winterMixed -> true

                smsAddress, smsTextPattern, smsText, sms -> true
                syncYandexDisk, sync, volume -> true

                else -> false
            }
        }
    }

    init {
        //        sharedPreferences.edit()
        //                .remove("last sync")
        //                .remove("full sync")
        //                .apply();

        sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String?) {
        if (LOG_ENABLED) {
            UtilsLog.d(TAG, "onSharedPreferenceChanged", "key == $key")
        }

        if (keys.isSyncKey(key)) {
            isChanged = true
        }
    }

    private fun getRevision(keyRevision: String?): Int {
        return sharedPreferences.getInt(keyRevision, -1)
    }

    private fun putRevision(keyRevision: String?, revision: Int) {
        sharedPreferences
            .edit()
            .putInt(keyRevision, revision)
            .apply()

        if (LOG_ENABLED) {
            UtilsLog.d(TAG, "putRevision", "$keyRevision == $revision")
        }
    }

    fun putFilterDate(dateFrom: Long, dateTo: Long) {
        sharedPreferences
            .edit()
            .putLong(keys.filterDateFrom, dateFrom)
            .putLong(keys.filterDateTo, dateTo)
            .apply()
    }

    fun putCalc(distance: String, cost: String, volume: String, cons: Int, season: Int) {
        sharedPreferences
            .edit()
            .putString(keys.distance, distance)
            .putString(keys.cost, cost)
            .putString(keys.volume, volume)
            .putInt(keys.consumption, cons)
            .putInt(keys.season, season)
            .apply()
    }

    private fun getString(key: String, defValue: String): String {
        return sharedPreferences.getString(key, defValue) ?: defValue
    }

    fun getString(key: String): String {
        return getString(key, "")
    }

    private fun getPreferences(preference: String?): ContentValues {
        val result = ContentValues()

        if (TextUtils.isEmpty(preference)) {
            for ((key, value) in sharedPreferences.all) {
                if (keys.isSyncKey(key)) {
                    when (value) {
                        is String -> result.put(key, value)
                        is Long -> result.put(key, value)
                        is Int -> result.put(key, value)
                        is Boolean -> result.put(key, value)
                        is Float -> result.put(key, value)

                        else -> throw IllegalArgumentException("$TAG -- getPreferences: unhandled class == ${value?.javaClass?.simpleName}")
                    }
                }
            }
        } else {
            when (preference) {
                keys.changed -> result.put(preference, isChanged)

                keys.databaseFullSync -> result.put(preference, isFullSync)

                keys.databaseRevision,
                keys.preferencesRevision -> result.put(preference, getRevision(preference))

                keys.lastSyncDateTime -> result.put(preference, lastSyncDateTime)
                keys.lastSyncHasError -> result.put(preference, lastSyncHasError)

                else -> throw IllegalArgumentException("$TAG -- getPreferences: unhandled preference == $preference")
            }
        }

        if (LOG_ENABLED) {
            for (key in result.keySet()) {
                UtilsLog.d(
                    TAG,
                    "getPreferences($preference)",
                    "key == $key, value == ${result.getAsString(key)}"
                )
            }
        }

        return result
    }

    fun getPreferencesCursor(preference: String? = null): Cursor {
        val matrixCursor = MatrixCursor(arrayOf("key", "value"))

        val preferences = getPreferences(preference)

        for (key in preferences.keySet()) {
            matrixCursor.addRow(arrayOf(key, preferences.get(key)))
        }

        return matrixCursor
    }

    @SuppressLint("ApplySharedPref")
    fun setPreferences(preferences: ContentValues?, preference: String?): Int {
        if (preferences == null || preferences.size() == 0) {
            return -1
        }

        if (LOG_ENABLED) {
            UtilsLog.d(
                TAG,
                "setPreferences",
                "preferences == $preferences, preference == $preference"
            )
        }

        if (TextUtils.isEmpty(preference)) {
            sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)

            val editor = sharedPreferences.edit()

            try {
                for ((key, value) in preferences.valueSet()) {
                    when (value) {
                        is String -> editor.putString(key, value)
                        is Long -> editor.putLong(key, value)
                        is Int -> editor.putInt(key, value)
                        is Boolean -> editor.putBoolean(key, value)
                        is Float -> editor.putFloat(key, value)

                        else -> throw IllegalArgumentException("$TAG -- setPreferences: unhandled class == ${value?.javaClass?.simpleName}")
                    }
                }
            } finally {
                editor.commit()

                sharedPreferences.registerOnSharedPreferenceChangeListener(this)
            }

            return preferences.size()
        } else {
            when (preference) {
                keys.changed -> isChanged = preferences.getAsBoolean(preference)

                keys.databaseRevision,
                keys.preferencesRevision -> putRevision(
                    preference,
                    preferences.getAsInteger(preference)
                )

                keys.databaseFullSync -> isFullSync = preferences.getAsBoolean(preference)

                keys.lastSyncDateTime -> lastSyncDateTime = preferences.getAsLong(preference)
                keys.lastSyncHasError -> lastSyncHasError = preferences.getAsBoolean(preference)

                else -> throw IllegalArgumentException("$TAG -- setPreferences: unhandled preference == $preference")
            }

            return 1
        }
    }

    @PreferenceType
    fun getPreferenceType(key: String): Int {
        return when (key) {
            keys.consumption,
            keys.season -> {
                PREFERENCE_TYPE_INT
            }

            keys.filterDateFrom, keys.filterDateTo,
            keys.mapCenterLatitude, keys.mapCenterLongitude -> {
                PREFERENCE_TYPE_LONG
            }

            else -> {
                PREFERENCE_TYPE_STRING
            }
        }
    }

    companion object {
        private const val TAG = "PreferencesHelper"

        @SuppressLint("StaticFieldLeak")
        private var instance: PreferencesHelper? = null

        private var LOG_ENABLED = false

        const val SYNC_NONE = java.lang.Long.MIN_VALUE

        const val PREFERENCE_TYPE_STRING = 0
        const val PREFERENCE_TYPE_INT = 1
        const val PREFERENCE_TYPE_LONG = 2

        @JvmStatic
        @Synchronized
        fun getInstance(context: Context): PreferencesHelper {
            if (instance == null) {
                instance = PreferencesHelper(context.applicationContext)
            }

            return instance as PreferencesHelper
        }
    }
}